pandas/core/arrays/numpy_.py

```diff
@@ -0,0 +1,427 @@
+from functools import wraps
+import numbers
+
+import numpy as np
+
+from pandas._libs import lib
+
+from pandas.core.dtypes.common import extract_array
+from pandas.core.dtypes.dtypes import ExtensionDtype
+from pandas.core.dtypes.generic import ABCIndexClass, ABCSeries
+from pandas.core.dtypes.inference import is_list_like
+
+from pandas import compat
+from pandas.core import nanops
+
+from .base import ExtensionArray, ExtensionOpsMixin
+
+
+def _validate_reduction_args(func):
+    """
+    Validate that unused keyword arguments are not set.
+    """
+    check = {'out', 'keepdims', 'initial', 'dtype'}
+
+    msg = (
+        "'{}' does not support the '{}' parameter."
+    )
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        if len(args) > 1:
+            raise TypeError(
+                "'{}' requires keyword arguments.".format(func.__name__)
+            )
+        for kw, value in kwargs.items():
+            if kw in check and value is not None:
+                raise TypeError(msg.format(func.__name__, kw))
+            elif kw == 'ovewrite_input' and kw is not False:
+                raise TypeError(msg.format(func.__name__, kw))
+
+        return func(*args, **kwargs)
+    return wrapper
+
+
+class PandasDtype(ExtensionDtype):
+    """
+    A Pandas ExtensionDtype for NumPy dtypes.
+
+    .. versionadded:: 0.24.0
+
+    This is mostly for internal compatibility, and is not especially
+    useful on its own.
+
+    Parameters
+    ----------
+    dtype : numpy.dtype
+    """
+    _metadata = ('_dtype',)
+
+    def __init__(self, dtype):
+        dtype = np.dtype(dtype)
+        self._dtype = dtype
+        self._name = dtype.name
+        self._type = dtype.type
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def type(self):
+        return self._type
+
+    @property
+    def _is_numeric(self):
+        # exclude object, str, unicode, void.
+        return self.kind in set('biufc')
+
+    @property
+    def _is_boolean(self):
+        return self.kind == 'b'
+
+    @classmethod
+    def construct_from_string(cls, string):
+        return cls(np.dtype(string))
+
+    def construct_array_type(cls):
+        return PandasArray
+
+    @property
+    def kind(self):
+        return self._dtype.kind
+
+    @property
+    def itemsize(self):
+        return self._dtype.itemsize
+
+
+try:
```

Comment:
Added a comment, and a note in the docstring.

```
# TODO(NumPy1.13): remove this
# Compat for NumPy 1.12, which doesn't provide NDArrayOperatorsMixin
# or __array_ufunc__, so those operations won't be available to people
# on older NumPys.
#
# We would normally write this as bases=(...), then "class Foo(*bases):
# but Python2 doesn't allow unpacking tuples in the class statement.
# So, we fall back to "object", to avoid writing a metaclass.
```